### ISSUE TITLE
Enhancement: [await-thenable] allow maybe-promises in promise aggregator checks

### ISSUE THREAD

**Issue by valadaptive** [typescript-eslint/typescript-eslint#11694]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

I debated whether I should report this as a bug in the `await-thenable` rule or as a proposal to move the behavior into a new rule.

The behavior in question is around the `await-thenable` rule's new functionality, where it will report an error if you pass an array of `Promise<T> | (some non-promise type)` into a promise aggregator method like `Promise.all`.

This was reported as a bug in https://github.com/typescript-eslint/typescript-eslint/issues/11609, since it rejects valid and useful code. That bug was "fixed" in https://github.com/typescript-eslint/typescript-eslint/pull/11611, but specifically only for the case of array *literals*. The following code is still rejected:

```ts
declare const mixedPromises: (Promise<number> | null)[];

Promise.all(mixedPromises);
```

I must admit I don't understand the logic behind #11611, or *why* exactly array literals are so special. However, this behavior seems to be deliberate, so I'm opening this as a feature request instead of a bug report.

I believe the "promise aggregator method must take all `Promise`s" check should never have been made part of `await-thenable`, because:

1. It has nothing to do with the `await` keyword, and doesn't even require the `await` keyword to be used in order to trigger. [The documentation for `await-thenable`](https://typescript-eslint.io/rules/await-thenable/) only mentions the `await` keyword, and doesn't mention promise aggregator methods anywhere.
2. It cannot be independently configured or disabled.
3. The `await-thenable` rule *permits* `await`ing unions of promises and non-promises:
    ```ts
    declare const mixedPromise: Promise<number> | null;
    async function foo() {
      // This is fine!
      await mixedPromise;
    }
    ```

I think that check should therefore be moved into a new rule. This new rule should allow unions of promises and non-promises by default, whether or not they're passed in as an array literal, and the stricter behavior should be an opt-in option (since it rejects code with valid use cases). I don't see any benefit to differentiating between array variables and array literals here--could @ronami perhaps fill me in on why they're treated differently right now?

### Fail Cases

```TypeScript
declare const nonPromises: number[];
Promise.all(nonPromises);

// With the "stricter" option enabled:
declare const mixedPromises: (Promise<number> | number)[];
Promise.all(mixedPromises);
```

### Pass Cases

```TypeScript
// With the "stricter" option *disabled*:
declare const mixedPromises: (Promise<number> | number)[];
Promise.all(mixedPromises);

declare const allPromises: Promise<number>[];
Promise.all(allPromises);
```

**Comment by kirkwaiblinger** [comment]
Hi @valadaptive!

1. Re documentation, yes, I agree that this feature warrants an example on the docs page.

1. I'm also sympathetic to the idea that calling `Promise.all(arrayOfMaybePromises)` is something that should be permitted (for same reasons as `await maybePromise`. So perhaps there is room for a change there. 

   ([playground with the code examples in this section](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBABRhwFssBnEACgEoAueC6ukAHlWSoBGIGAD4A3AChQkWAhTpseQsQBCHGvWZt1XXvyGjJ06HCRpMufEQwBhHZtbtKGnn0HDxEiWDy0M8AShgeABeJ056ADooCAgGAG1rcmcuZgAaeCgAdygsf2s1FIcM6zsixiYAXSZJKXATOXNFK2ISAFkoAE8hewrtcr13UXgAH3g3AyN62TMFS2UMFQ7ukF6tcJdBgxFR8f1hKZlTeQslUuWe8vXerY9diYOvH1Q-QhwcYLDe6NiEpIvVlcmCVVAC1sCFjYwUDqk9fP5AsBoREQCQYDAuqF4IlWuCMtlcvlVHjIeDKpJvjE4ojkS40RjOjV4AB6ZlIGL0cYgAhQbAANwQWEQ8AwAAs6PBgDgQLRUAByfxwAAOOBgGDhL3y7yRXUuKPpmLCONItNSEIKpuKkMtFXJEkpvwI2ptBsZYhZbMQHIQKtoeSwAvgQpF4tokulYeVqvVXhAAA8VWr4ABveAAXyAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOijdoTSODABfEPKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

   FWIW Array literals _are_ special, though. With something like this,
   ```ts
   declare function getAPromise(): Promise<number>;
   declare function getBPromise(): Promise<number>;
   declare function getCPromise(): Promise<number>;

   const bad = Promise.all([getAPromise(), await getBPromise(), getCPromise()]);
   ```
   
   we can see with our eyes that there's a mistake; `await getBPromise()` should just be `getBPromise()`. We can also analyze this with the linter and report on this case.

   But, something like this
   ```ts
   declare function getAMaybePromise(): Promise<number> | number;
   declare function getBMaybePromise(): Promise<number> | number;
   declare function getCMaybePromise(): Promise<number> | number;

   const good = Promise.all([getAMaybePromise(), getBMaybePromise(), getCMaybePromise()])
   ```
   is not a thing the rule should report, since each element _individually_ is `Promise<number> | number`. Whether or not that's an advisable API is also questionable, but it's not something that should be reported on for poor use of `Promise.all()`. So the rule does not report.

   On the other hand, with code such as the following,
   ```ts
   const badMaybePromiseArray = [getAPromise(), await getBPromise(), getCPromise()];
   Promise.all(badMaybePromiseArray); // false negative if this doesn't report

   const goodMaybePromiseArray = [getAMaybePromise(), getBMaybePromise(), getCMaybePromise()];
   Promise.all(goodMaybePromiseArray); // false positive if this does report
   ```

   The same mistake is present in the code in the first case, but it can't be told apart from the second case, since both simply have type `Array<Promise<number> | number>`. So, seeing an `Array<Promise<number> | number>` _may_ be defective. But if it is, the user _may or may not_ have a reasonable remedy.

   Plenty of lint rules _do reject perfectly valid code_ (which can always be remedied with an eslint-disable comment; don't be shy to use one when appropriate!). The difficulty is just in being judicious about whether to prefer false positives or false negatives in a particular tradeoff scenario. This maybe-promise array scenario here is exactly one such scenario where we are forced to accept a tradeoff, and the initial decision in https://github.com/typescript-eslint/typescript-eslint/pull/11267 / #11611 was to prefer false positives.

1. Re _naming_ (mentioned in https://github.com/typescript-eslint/typescript-eslint/issues/11609#issuecomment-3392987016), a related conversation came up here: https://github.com/typescript-eslint/typescript-eslint/issues/8858#issuecomment-2148991913. In short, the spirit of `await-thenable` is "do not await non-thenable things". This once only applied to _explicit_ use of `await`, but has naturally been extended to _implicit_ awaiting as well.

**Comment by valadaptive** [comment]
OK, I think I'm starting to understand? It seems there are two ways to use `Promise.all`:

- The first way is to call it with an array literal of statically-known size, with statically-known types (I believe TypeScript actually types it as a tuple). When using it like that, it makes sense to forbid mixing promises and non-promises--you know ahead of time *which* elements are non-promises, and can move them out of the array literal without changing the semantics or behavior of the code.
- The second way is to call it with a regular array that may be of arbitrary size and is defined somewhere else. This is a very common use case when you're doing any sort of "advanced" concurrency work. In *that* case, it's perfectly valid to mix non-promises into the array. For instance, if the array contains elements of type `Promise<T> | null`. I think there should not be any warning for this pattern.


I think instead of distinguishing by "array literal"/"not an array literal", it makes more sense to distinguish by "tuple"/"not tuple". This would allow detecting the code in your last example with a small tweak:

```ts
const badMaybePromiseArray = [getAPromise(), await getBPromise(), getCPromise()] as const; // Note the `as const`!
Promise.all(badMaybePromiseArray); // We can now infer that badMaybePromiseArray[1] is *never* a promise, and report here.
```

> 3. Re _naming_ (mentioned in [Bug: [await-thenable] inconsistent behavior on promise unions #11609 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/11609#issuecomment-3392987016)), a related conversation came up here: [Enhancement: [await-thenable] should check that `for-await` loop is used on an async iterable #8858 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/8858#issuecomment-2148991913). In short, the spirit of `await-thenable` is "do not await non-thenable things". This once only applied to _explicit_ use of `await`, but has naturally been extended to _implicit_ awaiting as well.

I don't see how #8858 is related to the naming complaint here. It involved `for await` loops, which clearly have an `await` keyword.

> Plenty of lint rules *do reject perfectly valid code* (which can always be remedied with an eslint-disable comment; don't be shy to use one when appropriate!). The difficulty is just in being judicious about whether to prefer false positives or false negatives in a particular tradeoff scenario.

My particular complaint here is twofold: the `Promise.all` behavior prefers false positives whereas every other `await-thenable` case prefers false negatives, and there is no way to independently configure it. I suppose I could live with keeping it as part of `await-thenable` if there was a way to independently configure it (*in addition* to better default behavior), but right now it feels like a bunch of only semi-related checks are being thrown into one lint rule with zero options.

When `await`ing a possibly-non-promise, the `await-thenable` rule *does not* trigger:
```ts
declare const maybePromise: number | Promise<number>;
async function foo() {
  await maybePromise; // This is OK
}
```

When `for await`ing a possibly-non-async iterable, the `await-thenable` rule likewise *does not* trigger:
```ts
declare const maybeAsyncIterable: Iterable<number> | AsyncIterable<number>;
async function foo() {
  for await (const bar of maybeAsyncIterable) { // This is OK too
    console.log(bar + 5);
  }
}
```

And yet, when `Promise.all`ing an *array* of possibly-non-promises, the `await-thenable` rule triggers:
```ts
declare const maybePromises: (number | Promise<number>)[];
async function foo() {
  await Promise.all(maybePromises); // This is not OK :(
}
```

**Comment by kirkwaiblinger** [comment]
First of all, I want to make sure I didn't give the wrong impression as to whether this proposal is being considered seriously. Note that there was even talk from 2 team members about adding a "mixed array" option during an early implementation attempt, see https://github.com/typescript-eslint/typescript-eslint/pull/8094#discussion_r1465004703. So proposing changes around the mixed arrays is definitely on solid ground! Just wanting to discuss to iron out the details.

> OK, I think I'm starting to understand? It seems there are two ways to use `Promise.all`:
> 
> * The first way is to call it with an array literal of statically-known size, with statically-known types (I believe TypeScript actually types it as a tuple). When using it like that, it makes sense to forbid mixing promises and non-promises--you know ahead of time _which_ elements are non-promises, and can move them out of the array literal without changing the semantics or behavior of the code.
> * The second way is to call it with a regular array that may be of arbitrary size and is defined somewhere else. This is a very common use case when you're doing any sort of "advanced" concurrency work. In _that_ case, it's perfectly valid to mix non-promises into the array. For instance, if the array contains elements of type `Promise<T> | null`. I think there should not be any warning for this pattern.
> 
> I think instead of distinguishing by "array literal"/"not an array literal", it makes more sense to distinguish by "tuple"/"not tuple". This would allow detecting the code in your last example with a small tweak:
> 
> const badMaybePromiseArray = [getAPromise(), await getBPromise(), getCPromise()] as const; // Note the `as const`!
> Promise.all(badMaybePromiseArray); // We can now infer that badMaybePromiseArray[1] is *never* a promise, and report here.

Yes and no... 

Tuples do share the property with array literals that we can determine whether entries each individually are maybe-promises or some entries are definitely-promises and some are definitely-not-promises, which in fairness is the main thing that I brought up above, for simplicity.

However, there are two other differences:

1. An array literal present the opportunity to flag the _particular_ element(s) in violation, whereas with an array/tuple passed in, the only choice is to flag the whole expression (often just an identifier). So, it's much more clear to the user _why_ it's being flagged, and it's a much less invasive change to move around the offending element(s), than to disassemble and possibly reassemble the expression [^1].

1. A tuple coming from some existing source indicates there is some importance to the order of its entries, so there's an argument for preserving that as there's an argument for preserving the order of a non-literal array passed in. _But_, an array literal provided to a promise aggregator just indicates you're passing in some values to be awaited in parallel that may well not have any meaningful ordering. 

(That is to say I think of a `Promise.all([...])` as semantically more like the following code)
```ts
// imagine Promise.all() used named properties rather than an array argument
const { foo, bar, baz } = await Promise.all({ foo: getFoo(), bar: getBar(), baz: getBaz() });
```

So, it is much more sensible IMO to rearrange the control flow to avoid non-promises when calling `Promise.all()` on an array literal than in other cases, even when they're tuples. ([here's a playground with some miscellaneous thoughts](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=C4TwDgpgBAQglgJ2ACwCYEMQBUCuYA20AvFANogToIBcUAdjgLYBGECANFIwPZ0q0MWbThhACmrBAF0A3ACg5AekVQ4qANZQA7shBQQ3HAHJUUVNygo4AZ07McwKKyra4KLhDmoIAY3xVoADMcOh9gOF5tCERUAEEwOAAKAEpackoaKAAFBG5GGwgAHkFJAD5OHj5kcSEOM0xaHLyC4ok2UtkFCAAPMG4kKHRrEFCoYNDwyIBzCGB4JDRMAEYUxtz86yL5lFFcAghSqABvOSgoZUs8Qigfbm8bnAQECD58PUD-KZnTZj0UaAQOGuAFpgU4HIMEFMcOhmG8oOo4HRTNxAlAcJtgvgAHSnKDPYCPOiDLToNzZdYFbHofD4RJaaIIOIJFLJeQAXy6vX6jiGIx8YxCYQixJmc0QO0wACZVhTmptCttFtgrgdjniLlhkDYzDYhpsWPD0Mj8ZRrAa4XpUZZkNBCftVNZtP1NgBCBRnC6MdDqaDWR7QAw4My8IyOYA%2Bu22zaDMBURxIm3QLS5OhTKD9bwId1nW50ayOdJUJobCAVXgoESYEsFKRQEgMmLxJJsqB4vMFsgUKhVkB1kjoUnkmubam0xJFhAjsv1EDTqRsjUqb2%2BqD%2B576QwhuhhqBgCH-GPMdA%2BTTAbhi20IVTE-7aVPpzNsHOmwkIYmT8tVXudTlyJHAGwgQntASqiAA8swABWviOCcZzdpkJTCHilT8PQbQcHiog1JIch-hc6BTGSxJgnmwAIHAABuECmD06CMPsXi%2BP4G7jMK0yzAAmhksrTq0tSlPI3h%2BAEgoTCKUBigAshWyCyshCDCSxYnsZMoqzAAIpgfGUgqilCVyfQDHyoxqZJYpgZgADMunylsErKpBMFhIc8HnCo1h5LM2pplAAD6Nj%2BWMNi6JwV7QKwPjoBiya6JCybQIOZLhH5AAGMlySkaUAPw5UuUBajq3CQPm6JOnglgWMMoTIKmhhOj0PgQGA6lOomGWzLJVTZTcMXWEi6YfNw6CpemYB6RA7XEp1wA8VQ2W4rmvCdp%2BXByT%2B9YkilcqlmOdKkGK80ICknCZT1yRnVpOnJAu8hnASRLHOt376BkvZQOyHJGTygw1QK5mceKCyiAALHZpaKo5EHQbBbkFUVTpUAgmBQPgbhsDSfXEqwYzPBA8I%2BMgxrfJChjIuF3AMjR14MtoxrADIqjAEYTr7s88KBP0GZ0BAwLhIwfOxUtHlQAA6n64S0lAxM0VVXCRmuAZJmjECBLyJqUVMyCOAN9zesAROcNatyPJsdgHsTLNOlBGIa22noqMadDcCAg1QAAjjgiCaNad7TvtKRQLEWQAJIix2haiJwiH9ttw6TYHh3XSAp1SdxvG3YujtQAAchYE3cHCECMK47j5N0tFrvyiimQKTUtW1IuPe%2Bz3R29PYveh50oEHX34T9Jn-eJHEacDkogAArBDLRWSAzlw%2BqOexPgXn0AXuTF6XWhuMgXBwJXpj-bXw8N61IrWNiDs3CtjhoXvJA9-J2c3-mUeYDHGRx8lCf2UnYraVTpddOc1M53TxC3YkRxZyf07vfT6HIgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQwHd3K78ALRE3YAjJOijdoTSODABfEPKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

In any case, my point is just, if we made a change to allow `Array<number | Promise<number>>` _unconditionally_ as a reasonable DevX concession, I think that tuples _should_ actually also come along for the ride. 

> > 3. Re _naming_ (mentioned in [Bug: [await-thenable] inconsistent behavior on promise unions #11609 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/11609#issuecomment-3392987016)), a related conversation came up here: [Enhancement: [await-thenable] should check that `for-await` loop is used on an async iterable #8858 (comment)](https://github.com/typescript-eslint/typescript-eslint/issues/8858#issuecomment-2148991913). In short, the spirit of `await-thenable` is "do not await non-thenable things". This once only applied to _explicit_ use of `await`, but has naturally been extended to _implicit_ awaiting as well.
> 
> I don't see how [#8858](https://github.com/typescript-eslint/typescript-eslint/issues/8858) is related to the naming complaint here. It involved `for await` loops, which clearly have an `await` keyword.

See the first comment, in which promise aggregators are specifically mentioned:
https://github.com/typescript-eslint/typescript-eslint/issues/8858#issuecomment-2040847895. 

[^1]: Well, you could do `Promise.all(myTuple.map(Promise.resolve))`, but that's a pretty yucky example of "coding to the linter" rather than just using a disable comment and writing code that makes sense.

**Comment by JoshuaKGoldberg** [comment]
@kirkwaiblinger would you say this is ready for a PR?

**Comment by kirkwaiblinger** [comment]
I think so... I guess the resolution would be that an option would allow for any input to be a mixed promise/non-promise array even if it's an array literal (provided that at least one entry is promise-y)

**Comment by xonaib** [comment]
I'd like to work on this. My plan is to add an option like allowForMaybePromiseArrays to await-thenable. When enabled, passing an array typed as (Promise<T> | NonPromise)[] to a promise aggregator would be allowed — consistent with how the rule already permits await (Promise<T> | NonPromise).

The change would be in the aggregator-checking logic, skipping the mixed-type check for non-literal arrays when the option is on. Happy to adjust the name or approach based on feedback before opening a PR.

**Comment by abouchard11** [comment]
@xonaib — apologies, I missed your comment here before I started work, and I've since opened #12889. Not trying to step on your toes: happy to close mine if you'd rather carry it forward, or to hand the branch over if that's easier.

On naming, I went with `allowMixedPromiseArrays` to follow the plugin's `allow…` convention and @kirkwaiblinger's "mixed promise/non-promise array" phrasing, but your `allowForMaybePromiseArrays` reads just as well and I have no attachment either way. One related point for whoever reviews: the option as implemented also covers tuples and non-array iterables (following the resolution in this thread), so a name that isn't array-specific may end up being more accurate.
